src/NITTA/LuaFrontend.hs

```diff
@@ -15,460 +11,397 @@ Copyright   : (c) Aleksandr Penskoi, 2019
 import qualified Data.HashMap.Strict as HM
-import Data.List (find, group, sort)
+import Data.Hashable
 import qualified Data.Map.Strict as M
 import Data.Maybe
 import Data.String
 import Data.String.ToString
-import qualified Data.String.Utils as S
-import Data.Text (Text, pack, unpack)
 import qualified Data.Text as T
 import Language.Lua
 import NITTA.Intermediate.DataFlow
 import qualified NITTA.Intermediate.Functions as F
-import NITTA.Utils (modify'_)
+import NITTA.Utils.Base
 import Text.Printf
 
+getUniqueLuaVariableName LuaValueInstance{luaValueInstanceName, luaValueInstanceIsConstant = True} luaValueAccessCount = "!" <> luaValueInstanceName <> "#" <> showText luaValueAccessCount
+getUniqueLuaVariableName LuaValueInstance{luaValueInstanceName, luaValueInstanceAssignCount} luaValueAccessCount
+    | T.head luaValueInstanceName == '_' = luaValueInstanceName
+    | otherwise = luaValueInstanceName <> "^" <> showText luaValueInstanceAssignCount <> "#" <> showText luaValueAccessCount
+
+data LuaStatement x = LuaStatement
+    { fIn :: [T.Text]
+    , fOut :: [LuaValueInstance]
+    , fName :: T.Text
+    , fValues :: [x]
+    , fInt :: [Int]
+    }
+    deriving (Show, Eq)
+
+-- | Stores information about a particular version of a variable. The version of a variable changes after assigning a new value to it.
+data LuaValueInstance = LuaValueInstance
+    { luaValueInstanceName :: T.Text
+    , luaValueInstanceAssignCount :: Int
+    , luaValueInstanceIsConstant :: Bool
+    }
+    deriving (Show, Eq)
+
 data FrontendResult v x = FrontendResult
     { frDataFlow :: DataFlowGraph v x
     , frTrace :: [TraceVar]
     , frPrettyLog :: [HM.HashMap v x] -> [HM.HashMap String String]
     }
 
-data TraceVar = TraceVar {tvFmt, tvVar :: Text}
+data TraceVar = TraceVar {tvFmt, tvVar :: T.Text}
     deriving (Show)
 
-defaultFmt = "%.3f"
-
--- |List contains IO Function names to be printed by default.
-listFuncIO = ["send", "recieve"]
-
-prettyLog traceVars hms = map prettyHM hms
-    where
-        prettyHM hm = HM.fromList $ map (fromMaybe undefined) $ filter isJust $ map prettyX $ HM.toList hm
-        prettyX (v0, x) = do
-            -- variables names end on #0, #1..., so we trim this suffix
-            let v = takeWhile (/= '#') $ toString v0
-            fmt <- v2fmt M.!? v
-            Just (toString v, printx (T.unpack fmt) x)
-        v2fmt = M.fromList $ map (\(TraceVar fmt v) -> (toString v, fmt)) traceVars
-        printx p x
-            | 'f' `elem` p = printf p (fromRational (toRational x) :: Double)
-            | 's' `elem` p = printf p $ show x
-            | otherwise = printf p x
+instance Hashable LuaValueInstance where
+    hashWithSalt i LuaValueInstance{luaValueInstanceName, luaValueInstanceAssignCount, luaValueInstanceIsConstant} =
+        ( (hashWithSalt i luaValueInstanceName * 31)
+            + hashWithSalt i luaValueInstanceAssignCount * 31
+        )
+            + hashWithSalt i luaValueInstanceIsConstant * 31
```

Comment:
Why 31?

Reply:
Unnecessary bracket?

Reply:
31 is a prime number, if we use it as a multiplication coefficient it gives more smooth distribution of a hashCode.
I have found it here, several years ago https://github.com/muhdkhokhar/test/blob/master/Joshua%20Bloch%20-%20Effective%20Java%20(3rd)%20-%202018.pdf (page 52)

but yes, I had some troubles here with brackets, fixed